### ISSUE TITLE
Issues regarding performance

### ISSUE THREAD

**Issue by konioy** [microsoft/LLM2CLIP#15]
I used the open-source model **LLM2CLIP-EVA02-L-14-336** to test the following images and sentences, but the results were very poor. Did I use it incorrectly? I didn't use xformers.
![Image](https://github.com/user-attachments/assets/6366fea2-28a7-413f-b72f-a036ec4e6072)
![Image](https://github.com/user-attachments/assets/28fae247-2de1-484c-acbd-2b3b78a02cf8)

captions = ["a student with a red hat"]
sim=[0.1643, 0.0739]

However, the results with model EVA02-L-14-336 were actually correct.
sim = [0.1539, 0.1591]

**Comment by konioy** [comment]
You can test it yourself with images and descriptions; similar descriptions might actually have lower similarity.

**Comment by konioy** [comment]
![Image](https://github.com/user-attachments/assets/dcf07e9a-707e-4ddb-9864-8b556e8d1ac0)
Did you make changes here? I don't see any difference.

**Comment by konioy** [comment]
I'm curious about what you changed. If it's the part below, it should have thrown an error before the modification, right?
![Image](https://github.com/user-attachments/assets/35e76129-d9c7-4c00-97c1-c855a8432824)

**Comment by konioy** [comment]
![Image](https://github.com/user-attachments/assets/9e3e2637-21f9-4af3-bc8b-b3bd7c1bbb86)
![Image](https://github.com/user-attachments/assets/10d2f9d1-f486-4855-a203-fbc44e525ddc)
captions = ["a deer"]

LLM2CLIP-EVA02-L-14-336 sim=[[0.3325, 0.2966]]
EVA02-L-14-336 sim = [[0.1321, 0.2197]]
The performance is still very poor.

**Comment by raytrun** [comment]
Please pull the latest code. I have tested it on the new example you provided, and here are the test results.
sim = [0.2690, 0.2939]

**Comment by konioy** [comment]
I have already updated the code under the llm2clip/eva_clip path, but the results are still incorrect.

**Comment by raytrun** [comment]
> I have already updated the code under the llm2clip/eva_clip path, but the results are still incorrect.

I am conducting the test on the commit ca5634faf. Please verify that the code version is consistent.

**Comment by konioy** [comment]
Yes

**Comment by raytrun** [comment]
Here is my test code, and this is the output result 
tensor([[0.2690, 0.2939]], device='cuda:0', dtype=torch.float16)
```python
import os

os.environ["CUDA_VISIBLE_DEVICES"] = "0"
import torch
from PIL import Image
from llm2vec import LLM2Vec
from eva_clip import create_model_and_transforms
from transformers import AutoModel, AutoConfig, AutoTokenizer

model, _, preprocess_val  = create_model_and_transforms('EVA02-CLIP-L-14-336', force_custom_clip=True)
ckpt = torch.load('LLM2CLIP-EVA02-L-14-336.pt')
model.load_state_dict(ckpt)
model = model.cuda().eval()

llm_model_name = 'microsoft/LLM2CLIP-Llama-3-8B-Instruct-CC-Finetuned'
config = AutoConfig.from_pretrained(
    llm_model_name, trust_remote_code=True
)

llm_model_name = 'microsoft/LLM2CLIP-Llama-3-8B-Instruct-CC-Finetuned'
config = AutoConfig.from_pretrained(
    llm_model_name, trust_remote_code=True
)
llm_model = AutoModel.from_pretrained(llm_model_name, torch_dtype=torch.bfloat16, config=config, trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained(llm_model_name)
llm_model.config._name_or_path = 'meta-llama/Meta-Llama-3-8B-Instruct' #  Workaround for LLM2VEC
l2v = LLM2Vec(llm_model, tokenizer, pooling_mode="mean", max_length=512, doc_max_length=512)

captions = ["a deer"]

image_path_1 = "im1.jpg"
image_path_2 = "im2.jpg"

image_1 = preprocess_val(Image.open(image_path_1)).cuda()
image_2 = preprocess_val(Image.open(image_path_2)).cuda()
image = torch.stack([image_1, image_2])

text_features = l2v.encode(captions, convert_to_tensor=True).to('cuda')

with torch.no_grad(), torch.cuda.amp.autocast():
    image_features = model.encode_image(image)
    text_features = model.encode_text(text_features)

    image_features /= image_features.norm(dim=-1, keepdim=True)
    text_features /= text_features.norm(dim=-1, keepdim=True)
    
print("sim:", text_features @ image_features.T)
```

**Comment by raytrun** [comment]
> Here is my test code, and this is the output result tensor([[0.2690, 0.2939]], device='cuda:0', dtype=torch.float16)
> 
> import os
> 
> os.environ["CUDA_VISIBLE_DEVICES"] = "0"
> import torch
> from PIL import Image
> from llm2vec import LLM2Vec
> from eva_clip import create_model_and_transforms
> from transformers import AutoModel, AutoConfig, AutoTokenizer
> 
> model, _, preprocess_val  = create_model_and_transforms('EVA02-CLIP-L-14-336', force_custom_clip=True)
> ckpt = torch.load('LLM2CLIP-EVA02-L-14-336.pt')
> model.load_state_dict(ckpt)
> model = model.cuda().eval()
> 
> llm_model_name = 'microsoft/LLM2CLIP-Llama-3-8B-Instruct-CC-Finetuned'
> config = AutoConfig.from_pretrained(
>     llm_model_name, trust_remote_code=True
> )
> 
> llm_model_name = 'microsoft/LLM2CLIP-Llama-3-8B-Instruct-CC-Finetuned'
> config = AutoConfig.from_pretrained(
>     llm_model_name, trust_remote_code=True
> )
> llm_model = AutoModel.from_pretrained(llm_model_name, torch_dtype=torch.bfloat16, config=config, trust_remote_code=True)
> tokenizer = AutoTokenizer.from_pretrained(llm_model_name)
> llm_model.config._name_or_path = 'meta-llama/Meta-Llama-3-8B-Instruct' #  Workaround for LLM2VEC
> l2v = LLM2Vec(llm_model, tokenizer, pooling_mode="mean", max_length=512, doc_max_length=512)
> 
> captions = ["a deer"]
> 
> image_path_1 = "im1.jpg"
> image_path_2 = "im2.jpg"
> 
> image_1 = preprocess_val(Image.open(image_path_1)).cuda()
> image_2 = preprocess_val(Image.open(image_path_2)).cuda()
> image = torch.stack([image_1, image_2])
> 
> text_features = l2v.encode(captions, convert_to_tensor=True).to('cuda')
> 
> with torch.no_grad(), torch.cuda.amp.autocast():
>     image_features = model.encode_image(image)
>     text_features = model.encode_text(text_features)
> 
>     image_features /= image_features.norm(dim=-1, keepdim=True)
>     text_features /= text_features.norm(dim=-1, keepdim=True)
>     
> print("sim:", text_features @ image_features.T)


Here are the two images you provided. If there are no unexpected issues, you should be able to obtain the same results.
![Image](https://github.com/user-attachments/assets/2f8b34b6-de52-4697-8b79-0e3ebc22a4fb)

**Comment by konioy** [comment]
I only changed trust_remote_code to False, and the result was incorrect.
[[0.3323, 0.2966]]

**Comment by raytrun** [comment]
Setting trust_remote_code=True is essential.

**Comment by konioy** [comment]
I have already downloaded the LLM2CLIP-Llama-3-8B-Instruct-CC-Finetuned model. What else should I do to run it offline?

**Comment by konioy** [comment]
![Image](https://github.com/user-attachments/assets/7bf95d63-87a2-4b04-8075-0495ad7c5f20)
Am I missing any files?

**Comment by raytrun** [comment]
> I have already downloaded the LLM2CLIP-Llama-3-8B-Instruct-CC-Finetuned model. What else should I do to run it offline?

The trust_remote_code=True flag is used for models whose code lives on the hub rather than natively in the Transformers library. 
The use of trust_remote_code does not conflict with offline operation.
After downloading the model, it can be executed offline.
